### PR TITLE
Issue 18/within range handler

### LINKED ISSUE #18: offset range lookup ("WITHIN"), get cells that are both LEFT and RIGHT
a common issue we have with databaker is that the api only supports absolute directions, so it can be `DIRECTLY` or `CLOSEST` in a given direction, but we have huge issues (and write a lot of nasty workarounds) for when the cells we want are in the middle (eg both LEFT and RIGHT, relative to different observations cells).

The solution would be a new keyword to search with an offset range.

Example:

```
HDim(cells, "Dimension1", WITHIN(-1, 1), ABOVE)
```

The -1 and 1 params above represent the offset to search, i.e one to the left through to one to the right, working upwards.

This pattern is illustrated, below, with the numbers showing the order which the lookup would search for a corresponding dimension value with these params.

|     | A | B | C | D | E | F | G | 
|----|----|----|----|----|----|----|----|
|     |     |    | 7    |  8  |  9   |    |    |
|    |     |    |  4    |  5  |  6  |     |    |
|    |     |    |   1   |   2 |  3  |    |     |
|   |     |    |      |  observation |  |   |

this would allow us to cleanly extract data from a table presenting the data with the following (and painfully common) pattern.

|     | A | B | C | D | E | F | G | 
|----|----|----|----|----|----|----|----|
|    |     |    |      |   |  |     |    |
|    |           |   2014   |         |     |          |  2015  |        |
|   |   obs  |   obs    |   obs |     |  obs | obs     | obs |

### PR DESCRIPTION
**Pull request by mikeAdamss**
Start of the WITHIN lookup engine. Draft PR with a few things still to do.

Things to do:
* read and play, take the time to get it first, this is not a trivial thing to understand.
* complete scenarios 3-8 in within.py, they should follow the same patterns already established with scenario 1 and 2 (do create more methods like those on 145 and 155, it's cleaner).
* tests, this will need a lot of tests
* test the FAILURES as well, we need this to be doing sane things if a DE passed in bad arguments, or specifies a range with no valid lookup in it.
* try it on a very large dataset, this may well need some refactoring for performance once it's confirmed as working (I'm not 100% that it'll have bad performance but we spent a lot of time on performance for the other engines, be silly to take a chance now). We don't necessarily need great performance right off the bat (though it'd be nice!) but we'll at least need to know if it needs looking at it in the future. 
* general neatening of things up, this is functional but very verbose, try and see if we can tighten up the code a bit anywhere ("no" is fine, but it's worth a pass or two and a few thoughts).

### CODE REVIEW

**Comment by CharlesRendle** [comment]
Ready for review. 

All scenarios completed with accompanying tests.
Have left the lookup functions in their verbose forms with the aim that this can be improved upon.

Please look into the within engine functions for searching up and down columns as I am not entirely happy with my logic behind them.
Also, I imagine there needs to be a wider breadth for the exception test scenarios. Any comments there would be great.

File: databaker/lookupengines/directly.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 from databaker.constants import ABOVE, BELOW, LEFT, RIGHT, DIRECTION_DICT
+from databaker.lookupengines.cell_val_override import cell_val_override
```
Comment (mikeAdamss) [inline databaker/lookupengines/directly.py]: Good thinking 👍 

though if we name this .py to something like `generic.py` we can put any other future shared functionality in it as well as cell value overrides``.

File: databaker/lookupengines/directly.py [hunk 2]
```diff
@@ -142,14 +143,15 @@ def lookup(self, cell):
...
                         # Apply str level cell value override if applicable
-                        if self.last_cell_found.value in self.cellvalueoverride:
-                            value = self.cellvalueoverride[self.last_cell_found.value]
-                        elif self.last_cell_found._cell in self.cellvalueoverride:
-                            value = self.cellvalueoverride[self.last_cell_found._cell]
-                        else:
-                            value = self.last_cell_found.value
-
-                        return self.last_cell_found, value
+                        #if self.last_cell_found.value in self.cellvalueoverride:
```
Comment (mikeAdamss) [inline databaker/lookupengines/directly.py]: well done. You can remove the commented out former approach code in all the engines when your override function is done.

File: databaker/lookupengines/closest.py [hunk 3]
```diff
@@ -235,11 +236,13 @@ def lookup(self, cell, index=None, ceiling=None, floor=0):
...
+            #if r["dimension_cell"].value in self.cellvalueoverride:
+            #    value = self.cellvalueoverride[r["dimension_cell"].value]
+            # Apply cell level cell value override if applicable
+            #elif r["dimension_cell"]._cell in self.cellvalueoverride:
+            #    value = self.cellvalueoverride[r["dimension_cell"]._cell]
+            #else:
+            #    value = r["dimension_cell"].value
+
+            #return r["dimension_cell"], value
+            return cell_val_override(r["dimension_cell"], value)
```
Comment (mikeAdamss) [inline databaker/lookupengines/closest.py]: would need to be `cell_val_override(r["dimension_cell"], self.cellvalueoverride)`

that `self.cellvalueoverride` is the dictionary of overrides that a DE has passed in, as it is that `value` variable doesn't exist.

I'd also recommend we switch to a pattern of:

```
cell, cell_value = cell_val_override(r["dimension_cell"], self.cellvalueoverride)
return cell, cell_value
```

its syntactic sugar that neither of us needs, but it'd make things a bit easier to follow for future engineers (I need to do similar elsewhere as I recall).

File: databaker/lookupengines/closest.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import json
 
 from databaker.constants import ABOVE, BELOW, LEFT, RIGHT, DIRECTION_DICT
+from databaker.lookupengines.cell_val_override import cell_val_override
```
Comment (mikeAdamss) [inline databaker/lookupengines/closest.py]: naming, see other comment.

File: databaker/lookupengines/directly.py [hunk 5]
```diff
@@ -142,14 +143,15 @@ def lookup(self, cell):
...
+                        #    value = self.cellvalueoverride[self.last_cell_found.value]
+                        #elif self.last_cell_found._cell in self.cellvalueoverride:
+                        #    value = self.cellvalueoverride[self.last_cell_found._cell]
+                        #else:
+                        #    value = self.last_cell_found.value
+
+                        #return self.last_cell_found, value
+                        return cell_val_override(self.last_cell_found, value)
```
Comment (mikeAdamss) [inline databaker/lookupengines/directly.py]: see comments in closest engine.

File: databaker/lookupengines/cell_val_override.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+def cell_val_override(found_cell, cellvalueoverride):
+#    # Apply str level cell value override if applicable
```
Comment (mikeAdamss) [inline databaker/lookupengines/cell_val_override.py]: JJ had an edge case this week where someone was applying both style of override to the same cell and it lead to some **very** confusing behaviour. It's the result of bad transform code primarily, but it makes sense to mitigate this while we're defining this function.

could we change this to (something like, feel free to twiddle):

```
import logging

def cell_val_override(found_cell, cellvalueoverride):

    str_override_applied = False
    cell_override_applied = False
    
    # Apply str level cell value override if applicable
    if found_cell.value in cellvalueoverride:
        str_override_applied = True
        value = cellvalueoverride[found_cell.value]

    # Apply cell level cell value override if applicable
    elif found_cell._cell in cellvalueoverride:
        cell_override_applied = True
        value = cellvalueoverride[found_cell._cell]

    # No overrides applied
    else:
        value = found_cell.value
        
    if str_override_applied and cell_overrides_applied:
        logging.warning('Both a str (cellvalueoverride=) and cell level (.AddCellValueOverride())'
                 f' override is being applied to cell {found_cell}. This should rarely or ever be necessary '
                 ' and can lead to confusing and difficult to debug behaviour".)

    return found_cell, value 
```
_note - written in browser, will likely need fixing_.

While I think the logic is correct do test this in anger please.

Comment (mikeAdamss) [reply databaker/lookupengines/cell_val_override.py]: it's a bit verbose, but I'd also maybe rename the function to `override_looked_up_cell` or `single_cell_override` or somesuch, again just a favour to future us.

File: databaker/lookupengines/within.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+
+        # Scenario 4: Scanning rightwards by row moving downwards from the top left of the table
+        elif self.direction == BELOW and self.direction_of_travel == RIGHT:
+            for (x_offset, y_offset) in self._xy_traveling_down_and_right():
+                potential_cell = [x for x in cell_bag if x.x == x_offset and x.y == y_offset]
+                if len(potential_cell) == 1:
+                    sequence.append(potential_cell[0])
+
+        # Scenario 5: Scanning upwards by column moving leftwards from the bottom right of the table
+        elif self.direction == ABOVE and self.direction_of_travel == LEFT:
```
Comment (mikeAdamss) [inline databaker/lookupengines/within.py]: needs to be `elif self.direction == LEFT and self.direction_of_travel == UP:`

This is the confusing bit where you have to switch your perspective. Imagine starring at a print out of the tab you're holding in front of your face - then rotate it 90 degrees right. This scenario is essentially mirroring the behaviour of scenario 2 but along a different axis.

You'll need to add more test data to the sheets (same as cats, dogs etc, but on the extreme left and right of our test table) to test this.

File: databaker/lookupengines/within.py [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+            for (x_offset, y_offset) in self._xy_traveling_up_cols_and_left():
+                potential_cell = [x for x in cell_bag if x.x == x_offset and x.y == y_offset]
+                if len(potential_cell) == 1:
+                    sequence.append(potential_cell[0])
+
+        # Scenario 6: Scanning downwards by column moving rightwards from the top right of the table
+        elif self.direction == BELOW and self.direction_of_travel == RIGHT:
```
Comment (mikeAdamss) [inline databaker/lookupengines/within.py]: `elif self.direction == RIGHT and self.direction_of_travel == DOWN:`

File: databaker/lookupengines/within.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+            for (x_offset, y_offset) in self._xy_traveling_down_cols_and_right():
+                potential_cell = [x for x in cell_bag if x.x == x_offset and x.y == y_offset]
+                if len(potential_cell) == 1:
+                    sequence.append(potential_cell[0])
+
+        # Scenario 7: Scanning upwards by column moving rightwards from the bottom left of the table
+        elif self.direction == ABOVE and self.direction_of_travel == RIGHT:
```
Comment (mikeAdamss) [inline databaker/lookupengines/within.py]: `elif self.direction == RIGHT and self.direction_of_travel == UP:`

File: databaker/lookupengines/within.py [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+            for (x_offset, y_offset) in self._xy_traveling_up_cols_and_right():
+                potential_cell = [x for x in cell_bag if x.x == x_offset and x.y == y_offset]
+                if len(potential_cell) == 1:
+                    sequence.append(potential_cell[0])
+
+        # Scenario 8: Scanning downwards by column moving leftwards from the top left of the table
+        elif self.direction == BELOW and self.direction_of_travel == LEFT:
```
Comment (mikeAdamss) [inline databaker/lookupengines/within.py]: `elif self.direction == LEFT and self.direction_of_travel == DOWN:`

File: databaker/lookupengines/within.py [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+            for (x_offset, y_offset) in self._xy_traveling_up_and_right():
+                potential_cell = [x for x in cell_bag if x.x == x_offset and x.y == y_offset]
+                if len(potential_cell) == 1:
+                    sequence.append(potential_cell[0])
+
+        # TODO - these
+        # Scenario 3: Scanning leftwards by row moving downwards from the top right of the table
+        elif self.direction == BELOW and self.direction_of_travel == LEFT:
+            for (x_offset, y_offset) in self._xy_traveling_down_and_left():
+                #print(x_offset)
```
Comment (mikeAdamss) [inline databaker/lookupengines/within.py]: don't forget to remove these (and any other temp comments) when you're done.

File: databaker/lookupengines/within.py (hunk 7 above)
Comment (mikeAdamss) [reply databaker/lookupengines/within.py]: I've added some comments against the scenario for this one

File: features/lookups.feature [hunk 12]
```diff
@@ -80,3 +80,153 @@ Scenario: Create a CONSTANT dimension
...
+    And we define the dimensions as
+        """
+        HDim(sheep_and_ducks, "Sheep and Ducks", WITHIN(left=1, right=2), BELOW)
+        """
+    Then the lookup from an observation in cell "C6" to the dimension "Sheep and Ducks" returns "{<E28 'Sheep'>}"
+    And the lookup from an observation in cell "F6" to the dimension "Sheep and Ducks" returns "{<E28 'Sheep'>}"
+    And the lookup from an observation in cell "G6" to the dimension "Sheep and Ducks" returns "{<I28 'Ducks'>}"
+    And the lookup from an observation in cell "I6" to the dimension "Sheep and Ducks" returns "{<I28 'Ducks'>}"
+
+#Scenario 5
```
Comment (mikeAdamss) [inline features/lookups.feature]: Assuming we puts cells of "Chips" and "Cheese" out on the left.

```
#Scenario 5
Scenario: Create a WITHIN LEFT, down to up columns dimensionsional lookup
    Given we load an xls file named "bakingtestdataset.xls"
    And select the sheet "Sheet1"
    And we define cell selections as
        | key               | value                                   |  
        | chips_and_cheese     | tab.excel_ref("<wherever you put them>").is_not_blank()       |
        | observations      | tab.excel_ref("C6:I25")                 |
    And we define the dimensions as
        """
        HDim(chips_and_cheese, "Chips and cheese", WITHIN(down=2, up=1), LEFT)
        """
    Then <steps go here, same pattern as usual>
 

```

File: features/lookups.feature [hunk 13]
```diff
@@ -80,3 +80,153 @@ Scenario: Create a CONSTANT dimension
...
+        """
+        HDim(cats_and_dogs, "Cats And Dogs", WITHIN(right=2, left=1), ABOVE)
+        """
+    Then the lookup from an observation in cell "C25" to the dimension "Cats And Dogs" returns "{<E3 'Cats'>}"
+    And the lookup from an observation in cell "F25" to the dimension "Cats And Dogs" returns "{<E3 'Cats'>}"
+    And the lookup from an observation in cell "G25" to the dimension "Cats And Dogs" returns "{<I3 'Dogs'>}"
+    And the lookup from an observation in cell "I25" to the dimension "Cats And Dogs" returns "{<I3 'Dogs'>}"
+
+#Scenario 6
+Scenario: Create a WITHIN BELOW, left to right columns dimensionsional lookup
```
Comment (mikeAdamss) [inline features/lookups.feature:170]: tweak along the lines of the scenario 5 comments

File: features/lookups.feature [hunk 14]
```diff
@@ -80,3 +80,153 @@ Scenario: Create a CONSTANT dimension
...
+        """
+        HDim(sheep_and_ducks, "Sheep and Ducks", WITHIN(left=1, right=2), BELOW)
+        """
+    Then the lookup from an observation in cell "C6" to the dimension "Sheep and Ducks" returns "{<E28 'Sheep'>}"
+    And the lookup from an observation in cell "F6" to the dimension "Sheep and Ducks" returns "{<E28 'Sheep'>}"
+    And the lookup from an observation in cell "G6" to the dimension "Sheep and Ducks" returns "{<I28 'Ducks'>}"
+    And the lookup from an observation in cell "I6" to the dimension "Sheep and Ducks" returns "{<I28 'Ducks'>}"
+
+#Scenario 7
+Scenario: Create a WITHIN ABOVE, left to right columns dimensionsional lookup
```
Comment (mikeAdamss) [inline features/lookups.feature:187]: tweak along the lines of the scenario 5 comments

File: features/lookups.feature [hunk 15]
```diff
@@ -80,3 +80,153 @@ Scenario: Create a CONSTANT dimension
...
+        """
+        HDim(cats_and_dogs, "Cats And Dogs", WITHIN(left=1, right=2), ABOVE)
+        """
+    Then the lookup from an observation in cell "C25" to the dimension "Cats And Dogs" returns "{<E3 'Cats'>}"
+    And the lookup from an observation in cell "F25" to the dimension "Cats And Dogs" returns "{<E3 'Cats'>}"
+    And the lookup from an observation in cell "G25" to the dimension "Cats And Dogs" returns "{<I3 'Dogs'>}"
+    And the lookup from an observation in cell "I25" to the dimension "Cats And Dogs" returns "{<I3 'Dogs'>}"
+
+#Scenario 8
+Scenario: Create a WITHIN BELOW, right to left columns dimensionsional lookup
```
Comment (mikeAdamss) [inline features/lookups.feature:204]: tweak along the lines of the scenario 5 comments

File: databaker/lookupengines/within.py [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+        # Pristine table bags are always exactly square
+        self.table_height = max([cell.y for cell in pristine_table])
+        self.table_width = max([cell.x for cell in pristine_table])
+
+        # TODO - we need to take a hard look at refactoring this for performance when its all working!
+        # I doubt this is even remotely efficient (so lets get the tests working, then we're free to rip the following to bits and make it faster)
+        sequence = []
+
+        if self.direction == ABOVE and self.direction_of_travel == LEFT:
+            for (x_offset, y_offset) in self._xy_traveling_up_and_left():
```
Comment (mikeAdamss) [inline databaker/lookupengines/within.py]: this is a refactoring comment (do these last) but in every scenario the code is the same: if 2 conditionals, run the loop with a specific method.

so something like (again example only):

```
def build_sequence(sequence_generator_method):
    """Sequence the cells appropriately based on the direction of travel and direction"""
    sequence = []
    for (x_offset, y_offset) in sequence_generator_method():
          potential_cell = [x for x in cell_bag if x.x == x_offset and x.y == y_offset]
          if len(potential_cell) == 1:
               sequence.append(potential_cell[0])
     return sequence
                    
  if self.direction == ABOVE and self.direction_of_travel == LEFT:
      sequence = build_sequence(self._xy_traveling_up_and_left())
  if self.direction == BELOW and self.direction_of_travel == RIGHT:
      sequence = build_sequence(self._xy_traveling_up_and_right())

File: databaker/lookupengines/within.py [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+            assert left is not None, msg2.format("right", "left")
+
+        if left is not None:
+            assert all(x is None for x in [above, below]), msg1.format("above and below", "left")
+            assert right is not None, msg2.format("left", "right")
+
+        # Get the direction of travel, this is dependant on the order in which the keyword args are passed
+        # e.g (left=1, right=2) is traveling RIGHT, while (right=1, left=1) is travelling LEFT
+        if left is not None:
+            # Consider the kwargs in the order they were recieved
```
Comment (mikeAdamss) [inline databaker/lookupengines/within.py:54]: refactoring (do this at the end), but we can  set `self.starting_offset` and `self.ending_offset` here and avoid doing it twice.

File: databaker/lookupengines/within.py [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+                    self.starting_offset = kwargs["left"]
+                    self.ending_offset = kwargs["right"]
+                    self.direction_of_travel = RIGHT
+                    break
+                if k == "right":
+                    self.starting_offset = kwargs["left"]
+                    self.ending_offset = kwargs["right"]
+                    self.direction_of_travel = LEFT
+                    break
+        elif above is not None:
```
Comment (mikeAdamss) [inline databaker/lookupengines/within.py:64]: see above comment

**Review by mikeAdamss — COMMENTED** [review]
have added some comments and snippets around the ABOVE, BELOW direction lookups (can have a session on this if you like, but see what makes sense to you first), as well as some refactoring stuff.

RE exceptions: the main thing is we want to tests to confirm we get an informative error if the DE does something silly, example: looking `WITHIN(left=2, right=1), ABOVE` when there **is no** dimension value relative to one of the observations with that instruction. I think we have the exception (line 359 in within.py at time of writing), we just need to confirm it triggers and make it something a bit less generic than ValueError.

**Pull request merged**